### PR TITLE
New package for RegCM, regional climate model.

### PR DESCRIPTION
**Pull request by bebosudo**
Hi,
I wrote a new package for RegCM, a regional climate model simulator software by ICTP: https://gforge.ictp.it/gf/project/regcm/

Any feedbacks are appreciated.

### CODE REVIEW

File: var/spack/repos/builtin/packages/regcm/package.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                          '-Wconversion', '-fimplicit-none', '-fbacktrace',
+                          '-ffree-line-length-0', '-finit-real=nan',
+                          '-ffpe-trap=zero,overflow,underflow', '-fcheck=all'])
+
+        elif name == 'ldlibs':
+            flags.extend(['-lnetcdff', '-lnetcdf'])
+            if self.compiler.fc.endswith('gfortran'):
+                flags.extend(['-lm', '-ldl'])
+            else:
+                flags.extend(['-lhdf5_hl', '-lhdf5', '-lz'])
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/regcm/package.py]: even though `hdf5` will be pulled in automatically for `netcdf`, if a package depends on another package it should be added as a direct dependency (even if it is also a transitive dependency).

**Comment by bebosudo** [comment]
I'm going to add some variants soon, so wait a sec before merging. :)

**Comment by alalazo** [comment]
@bebosudo Let me know when you are finished and I'll remove the tag

**Comment by bebosudo** [comment]
I'm done. I wanted to limit the number of processes in the make phase to 10/12 to avoid "race conditions", because sometimes the make crashes due to missing intermediate objects not already created.
I tried adding a `make_jobs = min(mp.cpu_count(), 12)` inside the class, but it's ignored. Any hint? Is it possible?

**Comment by alalazo** [comment]
> I wanted to limit the number of processes in the make phase to 10/12 to avoid "race conditions", because sometimes the make crashes due to missing intermediate objects not already created.

Hmm, this usually means that the `Makefile` of the package is not well written and does not specify dependencies appropriately. Imo you should add `parallel = False` as a class attribute rather than limiting the number of processes in the hope that timing will be good most of the times.

**Comment by bebosudo** [comment]
Okay, I forced the serial compilation. We can proceed go on with the PR.

File: var/spack/repos/builtin/packages/regcm/package.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            url='https://gforge.ictp.it/gf/download/frsrelease/259/1845/RegCM-4.7.0.tar.gz')
+
+    variant('debug', default=False,
+            description='Build RegCM using debug options.')
+    variant('profile', default=False,
+            description='Build RegCM using profiling options.')
+    variant('singleprecision', default=False,
+            description='Build RegCM using single precision float type.')
+
+    variant('knl', default=False,
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/regcm/package.py]: (FYI: no change required, but you may be interested) I assume that only one of these is set for a given architecture? These could be bundled into a single variant which enforces that only one can be chosen. Also: ideally, these could be determined automatically by examining `self.spec.architecture.target`, vs. requiring the user to specify the appropriate variant.

File: var/spack/repos/builtin/packages/regcm/package.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            description='Build RegCM using Intel Nehalem (SSE4.2) options.')
+
+    depends_on('netcdf')
+    depends_on('netcdf-fortran')
+    depends_on('hdf5')
+    depends_on('mpi')
+
+    # 'make' sometimes crashes when compiling with more than 10-12 cores.
+    # Moreover, parallel compile time is ~ 1m 30s, while serial is ~ 50s.
+    # The user can force the number of procs with: `spack install -j<N>`.
```
[the remark this answers is not in the document]
Comment (bebosudo) [reply var/spack/repos/builtin/packages/regcm/package.py]: Oh, didn't know! thanks, I'll remove line 40.

[the remark this answers is not in the document]
Comment (scheibelp) [reply var/spack/repos/builtin/packages/regcm/package.py]: Did you mean to push? I still see that line.

File: var/spack/repos/builtin/packages/regcm/package.py (hunk 2 above)
Comment (bebosudo) [reply var/spack/repos/builtin/packages/regcm/package.py]: Actually, if the compiler used is intel/pgi/others, the flags can be added together to create a fat binary for multiple archs, while on gcc it's not allowed, so the last set would be the only arch produced. I'm going to use a single variant to enforce a single architecture is built, if you agree.

Comment (bebosudo) [reply var/spack/repos/builtin/packages/regcm/package.py]: > these could be determined automatically by examining self.spec.architecture.target, vs. requiring the user to specify the appropriate variant.

When printing `self.spec.architecture` I get `linux-centos7-x86_64`, while `self.spec.architecture.target` returns `x86_64`. I tried digging in `self.spec.architecture`, but I found only generic architecture details.
These flags are meant to add instructions for specific Intel families, to take advantage of available extensions to the x86_64 architecture.
Moreover, the user may want to build from a different machine from the final one.
In any case, if not provided, a "classic" x86_64 executable is built, without any extension support.

So I'm pushing a new commit with a single multi-value variant for these extension flags.

File: var/spack/repos/builtin/packages/regcm/package.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bebosudo) [reply var/spack/repos/builtin/packages/regcm/package.py]: Fixed in a05c707.

**Comment by scheibelp** [comment]
Thanks!

**Pull request merged**